compose/rest/record.go

```diff
@@ -126,51 +128,163 @@ func (ctrl *Record) Read(ctx context.Context, r *request.RecordRead) (interface{
+	if rc.Values != nil {
+		rr := &types.Record{
+			NamespaceID: rc.NamespaceID,
+			ModuleID:    rc.ModuleID,
+			Values:      rc.Values,
+		}
+		batch = append(batch, &types.BulkRecordOperation{
+			Record:    rr,
+			Operation: types.OperationTypeCreate,
+		})
+	}
+
+	// Initialize sub records for creation, if defined
+	if rc.Records != nil {
+		for _, br := range rc.Records {
+			for _, rr := range br.Set {
+				// Default to current module and namespace; can be overwritten
+				if rr.NamespaceID == 0 {
+					rr.NamespaceID = rc.NamespaceID
+				}
+				if rr.ModuleID == 0 {
+					rr.ModuleID = rc.ModuleID
+				}
+				b := &types.BulkRecordOperation{
+					Record:    rr,
+					Operation: types.OperationTypeUpdate,
+				}
+
+				// If no RecordID is defined, we should create it
+				// It's ordered like this just for consistency with below Update(...)
+				if rr.ID == 0 {
+					b.Operation = types.OperationTypeCreate
+				}
+
+				// If a ref field is defined, then sub records should get their fields updated after
+				// parent record is created.
+				if br.RefField != "" {
+					if rc.Values == nil {
+						return nil, errors.New("missing parent record definition")
+					}
+
+					b.PreProcess = func(m, r *types.Record) (*types.Record, error) {
+						r.Values = r.Values.Set(&types.RecordValue{
+							Name:  br.RefField,
+							Value: strconv.FormatUint(m.ID, 10),
+						})
+						return r, nil
+					}
+				}
+
+				batch = append(batch, b)
+			}
+		}
+	}
+
+	rs, err := ctrl.record.With(ctx).Operate(batch...)
 
 	if rve, is := err.(*types.RecordValueErrorSet); is && !rve.IsValid() {
 		return ctrl.handleValidationError(rve), nil
 	}
 
-	return ctrl.makePayload(ctx, m, record, err)
+	if rc.Records == nil {
+		return ctrl.makePayload(ctx, m, rs[0], err)
+	}
+	return ctrl.makeFilterPayload(ctx, m, rs, nil, err)
 }
 
-func (ctrl *Record) Update(ctx context.Context, r *request.RecordUpdate) (interface{}, error) {
+func (ctrl *Record) Update(ctx context.Context, ru *request.RecordUpdate) (interface{}, error) {
 	var (
 		m   *types.Module
 		err error
 	)
 
-	if m, err = ctrl.module.With(ctx).FindByID(r.NamespaceID, r.ModuleID); err != nil {
+	if m, err = ctrl.module.With(ctx).FindByID(ru.NamespaceID, ru.ModuleID); err != nil {
 		return nil, err
 	}
 
-	record, err := ctrl.record.With(ctx).Update(&types.Record{
-		ID:          r.RecordID,
-		NamespaceID: r.NamespaceID,
-		ModuleID:    r.ModuleID,
-		Values:      r.Values,
-	})
+	batch := make([]*types.BulkRecordOperation, 0)
+
+	// Initialize parent record for creation, if defined
+	if ru.Values != nil {
+		rr := &types.Record{
+			ID:          ru.RecordID,
+			NamespaceID: ru.NamespaceID,
+			ModuleID:    ru.ModuleID,
+			Values:      ru.Values,
+		}
+		batch = append(batch, &types.BulkRecordOperation{
+			Record:    rr,
+			Operation: types.OperationTypeUpdate,
+		})
+	}
+
+	// Initialize sub records for creation, if defined
+	if ru.Records != nil {
+		for _, br := range ru.Records {
```

Comment:
should/could this logic be part of `BulkRecordSet` (not as generated func, as additional custom on set)?